Photolysis constant widget

RedHorizonss opened the pull request:
Added Photolysis graph and linked the above graph to the photolysis graph.

danielhollas commented:
Hi @RedHorizonss,

I've tried to test your branch locally, but when I loaded one of my worflows the application crashed with an exception. Please see the full traceback in #167. Could you or one of your teammates please take a look? Also please cleanup this PR a bit by removing the various temporary files that you used in the development.

I've also had a brief look at the code itself and in general it looks great! :clap: I also appreciate that you added docstrings, I've been much less consistent with them myself. I will have a furtrher look once I am able to test the code.

RedHorizonss commented:
Hey @danielhollas ,

The issue that caused the application to crash was that the CH4 molecule that would initially load had a cross section range that did not fit the actinic flux range. So we've added a mask to combat this issue and cleaned up the extra development files. Hopefully it should be better now!

danielhollas commented on aiidalab_ispg/spectrum_analysis.py:
```diff
@@ -224,3 +241,389 @@ def _observe_disabled(self, change):
+
+        if self.spectrum_data is None:
+            return
+        if plot_type == "LOW":
+            j_values = self.calculation(1, quantum_yield=quantumY)
+            wavelengths = self.flux_data[0]
+            self.plot_line(wavelengths, j_values, label="label")
+            self.add_log_axis(wavelengths, 1, label="log")
```
There is a lot of repetition here. I would propose the `self.flux_data` to be a dictionary, and then you can use the `plot_type` as a key, and just pass it down to the `plot_line` and `add_log_axis` functions. 
Also the strings "LOW", "HIGH" and "MED" are arbitrarily defined in multiple places which is a bit error prone. It would be nicer to use the Enum type for this, in the same way as I use it in spectrum.py for other such variables. 

However, I would propose to do this refactoring in a separate followup PR.

danielhollas commented on aiidalab_ispg/spectrum_analysis.py:
```diff
@@ -224,3 +241,389 @@ def _observe_disabled(self, change):
+
+    @traitlets.observe("disabled")
+    def _observe_disabled(self, change):
+        """
+        Observe changes to the "disabled" trait and update the state of the flux_toggle and yield_slider widgets accordingly.
+
+        :param change: A dictionary containing information about the change to the "disabled" trait.
+        """
```
Because the observer pattern is so common in all the widgets, here I'd say the docstring is a bit redundant, especially since it is a private method anyway. The function name and the decorator already make things clear enough.

danielhollas commented on aiidalab_ispg/spectrum_analysis.py:
```diff
@@ -224,3 +241,389 @@ def _observe_disabled(self, change):
+        mol_intensity_interp = np.interp(self.flux_data[0], mol_wlength, mol_intensity)
+        return mol_intensity_interp
+
+    def process(self, file, delimiter):
```
Looks like this function is not needed.

danielhollas commented on aiidalab_ispg/spectrum_analysis.py:
```diff
@@ -224,3 +241,389 @@ def _observe_disabled(self, change):
+        :return: The interpolated cross section data.
+        :rtype: np.ndarray
+        """
+        mol_wlength, xsection = self.spectrum_data
+        mol_wlength = np.flip(mol_wlength)
+        mol_intensity = xsection
+        mol_intensity = np.flip(mol_intensity)
+        wl_max = mol_wlength.max()
+        mol_wlength, mol_intensity = self.mask_data(
+            mol_wlength, mol_intensity, 280, np.floor(wl_max)
+        )
+        mol_intensity_interp = np.interp(self.flux_data[0], mol_wlength, mol_intensity)
+        return mol_intensity_interp
```
Just suggesting renaming variables to make things clearer. No need to use confusing abbreviations.
Also, the magic number 280 needs to be defined somewhere as a (constant) variable, perhaps `self.min_flux_wavelength` or something? 
Though I think the interpolation range should not be hardcoded, but should be computed based on the spectral range of the cross section. Something like

x_min = max(min(wavelength), 280)

```suggestion
        wavelengths, cross_section = self.spectrum_data
        x = np.flip(wavelengths)
        y = np.flip(cross_section)
        x_max = wavelengths.max()
        x_masked, y_masked = self.mask_data(x, y, 280, np.floor(x_max))

        cross_section_interpolated = np.interp(self.flux_data[0], x_masked, y_masked)
        return cross_section_interpolated
```

danielhollas commented on aiidalab_ispg/spectrum_analysis.py:
```diff
@@ -224,3 +241,389 @@ def _observe_disabled(self, change):
+        data = np.asarray(data[1:]).transpose()
+        return data.astype(float)
+
+    def mask_data(self, array_wlength, array_intensities, minimum, maximum):
```
It's too late for me to understand this function, I'll do it in the second pass. :-D

danielhollas commented on aiidalab_ispg/spectrum_analysis.py:
```diff
@@ -224,3 +241,389 @@ def _observe_disabled(self, change):
+        spectrum_max = max(self.spectrum_data[0])
+
+        # Check end of spectrum data overlaps with flux data
+        if spectrum_max >= flux_min:
```
In principle, you should also check that the other side of the interval to be on the safe side. :-) 

```suggestion
        if spectrum_max >= flux_min and spectrum_min < flux_max:
```

danielhollas reviewed (changes requested):
Hi folks,

thanks so much for all you work on this, it looks great! 

Sorry for dragging my feet on the review (this is what happens with big PRs :D ). The comments are made are mostly minor. If I suggested somewhere a bigger refactor, it's perhaps better for a follow-up PR. Because this is a large change, we might need a couple of iterations to converge, but that is fine.

One thing that is important here is to make sure that the x-axis range is set correctly, based on both the flux range and the calculated cross section range. Perhaps this is already done correctly, but I haven't fully verified it. Please double check that, this is important to get right since an error here would produce errorneous total J values.

Thanks again! If some comments are not clear, please comment here on Github, not via email.

kon-218 commented on aiidalab_ispg/spectrum_analysis.py:
```diff
@@ -224,3 +241,389 @@ def _observe_disabled(self, change):
+            value="HIGH",
+        )
+
+        self.flux_toggle.observe(self._observe_flux_toggle, names="value")
```
Hi daniel could you expand on this a little? do you mean with traitlets.observe() before the oberve_flux_toggle function?

danielhollas replied:
Yes, exactly.

danielhollas replied:
Never mind, it seems that the decorator approach would not work since the traitlet is defined in the constructor. Sorry for confusion.

danielhollas commented on aiidalab_ispg/spectrum.py:
```diff
@@ -507,6 +521,23 @@ def _plot_spectrum(
+            total_cross_section_nm = np.zeros(Spectrum.N_SAMPLE_POINTS)
+
+            for conf_id, conformer in enumerate(self.conformer_transitions):
+                spec = Spectrum(conformer["transitions"], conformer["nsample"])
```
I don't think we need to create a new instance here, we can reuse the one from above.
```suggestion
```

danielhollas commented on aiidalab_ispg/spectrum_analysis.py:
```diff
@@ -224,3 +241,389 @@ def _observe_disabled(self, change):
+        :rtype: tuple
+        """
+        z = np.loadtxt(
+            fname="/home/jovyan/apps/aiidalab-ispg/aiidalab_ispg/static/StandardActinicFluxes2.csv",
```
Resolved in #195

danielhollas commented:
Hi folks,

I've merged your work a while back in #195. Thank you again! We've been presenting ATMOSPEC with Basile on multiple occasions over past few months and got great feedback, and your widget has definitely been a cherry on top in our demonstrations. :smiley: 

Those of you who pushed commits to this PR are immortalized in the [merge commit](https://github.com/ispg-group/aiidalab-ispg/commit/cdc38fb4b65827302bc86294fcfe6f5c88f92dda) and are listed as [contributors on Github](https://github.com/ispg-group/aiidalab-ispg/graphs/contributors). I have also added all of your names at the top of the file with the widget, and you also star in credits at the end of the[ screencast that we made](https://youtu.be/1ePj1hhOFdw). :-)  

CC @RedHorizonss @kon-218 @WillHobson
